feat: use 'formatlistpat' length for breakindent

habamax opened the pull request:
Provided 'set briopt+=list:-1' use length of the matched 'formatlistpat' as additional indent to align contents of the list item.

@chrisbra pls check it, related to your PR #8564 

PS, not sure how to write a proper test though.

habamax committed:
feat: use 'formatlistpat' length for breakindent

Provided 'set briopt+=list:-1' use length of the matched 'formatlistpat'
as additional indent to align contents of the list item.

chrisbra commented on src/indent.c:
```diff
@@ -958,7 +958,10 @@ get_breakindent_win(
+		if (wp->w_briopt_list > 0)
+		    bri += wp->w_briopt_list;
+		else
+		    bri += (*regmatch.endp - *regmatch.startp);
```
> bri += (*regmatch.endp - *regmatch.startp);

I think you want to use `bri = ...`, not adding to the currently calculated indent.

habamax replied:
what about showbreak and other stuff then? let me check it...

chrisbra replied:
yeah, likely this might be needed for some options, and for other not (like `shift:`) :(

habamax replied:
If you use `bri = ` then `set showbreak=>` interferes:
![image](https://user-images.githubusercontent.com/234774/126285099-62414dcf-06f9-4ef1-a357-d0204645e9c2.png)

habamax replied:
> yeah, likely this might be needed for some options, and for other not (like `shift:`) :(

With current implementation in this PR it works ok with or without showbreak:
![image](https://user-images.githubusercontent.com/234774/126285727-f42c1fbf-bbc2-4c8e-b36d-c6ffd52ca587.png)

habamax replied:
I take it back, it should not add to bri...

habamax replied:
Fixed with showbreak char in mind.

habamax committed:
fix: do not add indent when 'formatlistpat' len is used

Otherwise it looks like:

    * hello world
         next line has too much indent

Showbreak len is substracted after indent for the list is
calculated for proper alignment.

habamax committed:
chore: add formatlistpat test

habamax committed:
fix: use local formatlistpat

chrisbra commented on src/indent.c:
```diff
@@ -941,15 +941,11 @@ get_breakindent_win(
 
     // add additional indent for numbered lists
-    if (wp->w_briopt_list > 0)
+    if (wp->w_briopt_list > 0 || wp->w_briopt_list == -1)
```
> if (wp->w_briopt_list > 0 || wp->w_briopt_list == -1)

if (wp->w_briopt_list != 0)

chrisbra commented on src/testdir/test_breakindent.vim:
```diff
@@ -759,6 +759,23 @@ func Test_breakindent20_list()
+  call s:compare_lines(expect, lines)
   " reset linebreak option
   " Note: it indents by one additional
   " space, because of the leading space.
```
please add another test with different levels of lists, so like this:
```
   * foobar with
     long text
   *** some other item
       with a lot of text
```

chrisbra replied:
also test with `briopt+=sbr:1`  or `briopt+=shift:2`

habamax replied:
It would be a bit more tricky -- I don't know how to start tests locally :)
Will add

chrisbra replied:
from the testdir, you can simply run `make test_breakindent.res` and then inspect the file messages (or whatever the logfile is named, I forgot :/)

habamax committed:
style: add braces, simplify condition

habamax committed:
test: add multilevel list test

habamax committed:
test: fix tests

Now with local test run first.

habamax committed:
test: add showbreak test

habamax committed:
test: add shift:2 test

habamax committed:
test: make tested options more explicit

With setl briopt+=list:-1 we get double list opts in the setting.
